src/Lean/Meta/Tactic/Rewrites.lean

```diff
@@ -0,0 +1,378 @@
+It would be good to solve this problem!
+
+Implementation: we reverse the results from `getMatch`,
+so that we return lemmas matching larger subexpressions first,
+and amongst those we return more specific lemmas first.
+-/
+partial def getSubexpressionMatches (op : Expr → MetaM (Array α)) (e : Expr) : MetaM (Array α) := do
+  match e with
+  | .bvar _ => return #[]
+  | .forallE _ _ _ _ =>
+    forallTelescope e fun args body => do
+      args.foldlM (fun acc arg => return acc ++ (← getSubexpressionMatches op (← inferType arg)))
+        (← getSubexpressionMatches op body).reverse
+  | .lam _ _ _ _
+  | .letE _ _ _ _ _ =>
+    lambdaLetTelescope e (fun args body => do
+      args.foldlM (fun acc arg => return acc ++ (← getSubexpressionMatches op (← inferType arg)))
+        (← getSubexpressionMatches op body).reverse)
+  | _ =>
+    let init := ((← op e).reverse)
+    e.foldlM (init := init) (fun a f => return a ++ (← getSubexpressionMatches op f))
+
+/--
+Find lemmas which can rewrite the goal.
+
+This core function returns a monadic list, to allow the caller to decide how long to search.
+See also `rewrites` for a more convenient interface.
+-/
+-- We need to supply the current `MetavarContext` (which will be reused for each lemma application)
+-- because `MLList.squash` executes lazily,
+-- so there is no opportunity for `← getMCtx` to record the context at the call site.
+def rewriteCandidates (hyps : Array (Expr × Bool × Nat))
+    (moduleRef : LazyDiscrTree.ModuleDiscrTreeRef (Name × Bool × Nat))
+    (target : Expr)
+    (forbidden : NameSet := ∅) :
+    MetaM (Array ((Expr ⊕ Name) × Bool × Nat)) := do
+  -- Get all lemmas which could match some subexpression
+  let candidates ← getSubexpressionMatches (rwFindDecls moduleRef) target
+  -- Sort them by our preferring weighting
+  -- (length of discriminant key, doubled for the forward implication)
+  let candidates := candidates.insertionSort fun (_, _, rp) (_, _, sp) => rp > sp
+
+  -- Now deduplicate. We can't use `Array.deduplicateSorted` as we haven't completely sorted,
+  -- and in fact want to keep some of the residual ordering from the discrimination tree.
+  let mut forward : NameSet := ∅
+  let mut backward : NameSet := ∅
+  let mut deduped := #[]
+  for (l, s, w) in candidates do
+    if forbidden.contains l then continue
+    if s then
+      if ¬ backward.contains l then
+        deduped := deduped.push (l, s, w)
+        backward := backward.insert l
+    else
+      if ¬ forward.contains l then
+        deduped := deduped.push (l, s, w)
+        forward := forward.insert l
+
+  trace[Tactic.rewrites.lemmas] m!"Candidate rewrite lemmas:\n{deduped}"
+
+  -- Lift to a monadic list, so the caller can decide how much of the computation to run.
+  let hyps := hyps.map fun ⟨hyp, symm, weight⟩ => (Sum.inl hyp, symm, weight)
+  let lemmas := deduped.map fun ⟨lem, symm, weight⟩ => (Sum.inr lem, symm, weight)
+  pure <| hyps ++ lemmas
+
+structure FinRwResult where
```

Comment:
What is the `Fin` in the name referring to?

Reply:
This was a bad refactor